### PR TITLE
Sort discovered dependency files

### PR DESCRIPTION
**Pull request by neonichu**
Sort files which are less likely to be authored at the end, in order to find cycles based on authored files first

See <rdar://problem/37768597>

### CODE REVIEW

**Comment by neonichu** [comment]
@swift-ci please smoke test

File: lib/Core/BuildEngine.cpp [hunk 1]
```diff
@@ -1420,7 +1425,12 @@ class BuildEngineImpl {
     }
 
     auto dependencyID = getKeyID(key);
-    taskInfo->discoveredDependencies.push_back(dependencyID);
+    // Sort files which are less likely to be authored at the end, in order to find cycles based on authored files first.
+    if (ends_with(key, "module.modulemap")) {
```
Comment (ddunbar) [inline lib/Core/BuildEngine.cpp]: StringRef should have a method like this on it that you can just use instead (with basically no performance penalty).

**Comment by neonichu** [comment]
Added a test:

Before the change the cycle would be reported as

```
<unknown>:0: error: cycle detected while building: target '' -> node 'output' -> command 'output' -> node 'output-1' -> command 'output-1' -> node 'module.modulemap' -> command 'output-2' -> node 'output-1'
```

After the change the cycle will be reported as

```
<unknown>:0: error: cycle detected while building: target '' -> node 'output' -> command 'output' -> node 'output-1' -> command 'output-1' -> node 'header-1' -> command 'output-2' -> node 'output-1'
```

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Pull request merged**